installers/binstall/src/system/windows.rs

```diff
@@ -1,5 +1,169 @@
+// most of this code is taken wholesale from this:
+// https://github.com/rust-lang/rustup/blob/master/src/cli/self_update/windows.rs
 use crate::{Installer, InstallerError};
 
+use std::io;
+
+/// Adds the downloaded binary in Installer to a Windows PATH
 pub fn add_binary_to_path(installer: &Installer) -> Result<(), InstallerError> {
-    unimplemented!()
+    let windows_path = get_windows_path_var()?;
+    let bin_path = installer
+        .get_bin_dir_path()?
+        .to_str()
+        .ok_or(InstallerError::PathNotUnicode)?
+        .to_string();
+    if let Some(old_path) = windows_path {
+        if let Some(new_path) = add_to_path(&old_path, &bin_path) {
+            apply_new_path(&new_path)?;
+        }
+    }
+
+    Ok(())
+}
+
+// ---------------------------------------------
+// https://en.wikipedia.org/wiki/Here_be_dragons
+// ---------------------------------------------
+// nothing is sacred beyond this point.
+
+// Get the windows PATH variable out of the registry as a String. If
+// this returns None then the PATH variable is not unicode and we
+// should not mess with it.
+fn get_windows_path_var() -> Result<Option<String>, InstallerError> {
+    use winreg::enums::{HKEY_CURRENT_USER, KEY_READ, KEY_WRITE};
+    use winreg::RegKey;
+
+    let root = RegKey::predef(HKEY_CURRENT_USER);
+    let environment = root.open_subkey_with_flags("Environment", KEY_READ | KEY_WRITE)?;
+
+    let reg_value = environment.get_raw_value("PATH");
+    match reg_value {
+        Ok(val) => {
+            if let Some(s) = string_from_winreg_value(&val) {
+                Ok(Some(s))
+            } else {
+                tracing::warn!("the registry key HKEY_CURRENT_USER\\Environment\\PATH does not contain valid Unicode. \
+                       Not modifying the PATH variable");
+                Ok(None)
+            }
+        }
+        Err(ref e) if e.kind() == io::ErrorKind::NotFound => Ok(Some(String::new())),
+        Err(e) => Err(e)?,
+    }
+}
+
+// This is used to decode the value of HKCU\Environment\PATH. If that
+// key is not unicode (or not REG_SZ | REG_EXPAND_SZ) then this
+// returns None.  The winreg library itself does a lossy unicode
+// conversion.
+fn string_from_winreg_value(val: &winreg::RegValue) -> Option<String> {
+    use std::slice;
+    use winreg::enums::RegType;
+
+    match val.vtype {
+        RegType::REG_SZ | RegType::REG_EXPAND_SZ => {
+            // Copied from winreg
+            let words = unsafe {
+                #[allow(clippy::cast_ptr_alignment)]
+                slice::from_raw_parts(val.bytes.as_ptr().cast::<u16>(), val.bytes.len() / 2)
```

Comment:
oh boy 😂

Reply:
yyyeeaaaahhhh - this is taken right from rustup lol so it should be ok but.... damn. windowwwwss